ui/app/main.js

```diff
@@ -1,82 +1,131 @@
+                            renderConstituentData(constituent);
+                        });
+                } else {
+                    renderLoginOptions();
                 }
-
-                /**
-                 *  Access token is valid. Fetch constituent record.
-                 */
-                $http.get('/api/constituents/280').then(function (res) {
-                    $scope.constituent = res.data;
-                    $scope.isReady = true;
-                });
             });
+    }
 
-            /**
-             * Opens a new popup window and redirects it to our login route.
-             * After a successful login within the popup, the popup is redirected to the `#/auth-success`
-             * route, which closes the popup window and returns focus to the parent window.
-             */
-            $scope.popupLogin = function () {
-                var popup;
+    // Render login options
+    function renderLoginOptions() {
+        const loginOptions = `
+          <h1>SKY API Authorization Code Flow Tutorial</h1>
+          <div class="well">
+            <h4 class="well-title">Applications implementing the authorization code flow may be authorized in one of two ways:</h4>
+            <div class="row">
+              <div class="col-sm-6">
+                <div class="panel panel-default panel-body">
+                  <p>
+                    <a href="/auth/login" class="btn btn-primary btn-block btn-lg"><i class="fa fa-external-link"></i> Authorize using redirect</a>
+                  </p>
+                  <p>
+                    The user authorizes the application after being redirected to the Blackbaud Authorization website.
+                  </p>
+                </div>
+              </div>
+              <div class="col-sm-6">
+                <div class="panel panel-default panel-body">
+                  <p>
+                    <button id="popup-login" class="btn btn-primary btn-block btn-lg" target="login-iframe"><i class="fa fa-window-restore"></i> Authorize using popup</button>
+                  </p>
+                  <p>
+                    The user authorizes the application using a browser popup window.
+                  </p>
+                </div>
+              </div>
+            </div>
+          </div>
+        `;
+        app.innerHTML = loginOptions;
 
-                popup = $window.open('auth/login?redirect=/%23/auth-success', 'login', 'height=450,width=600');
+        // Add event listener to popup login button
+        document.getElementById("popup-login").addEventListener("click", function () {
+            const popup = window.open('auth/login?redirect=/%23/auth-success', 'login', 'height=450,width=600');
+            if (window.focus) {
+                popup.focus();
+            }
 
-                if ($window.focus) {
-                    popup.focus();
-                }
-            };
+            let intervalId = setInterval(function () {
```

Comment:
This setInterval should also check whether `popup` has already been closed. In testing, I can close the popup and the `fetch('/auth/authenticated')` call will still fire every 200ms. In addition to checking whether the user is authorized, if `popup` is closed, `clearInterval(intervalId)` should also be called.


-----
Bonus: In testing, I can launch multiple popups by clicking the "Authorize using popup" more than once, which will set up multiple redundant `intervals`. Ideally, this page would keep track of the popup and only launch one pop at a time--along with only having one `interval` set at a time. If a popup exists, the page could apply `focus` to that popup the next time the "Authorize using popup" is clicked. However, for demonstration purposes, I don't think this is necessary--only a nice to have.

Reply:
Added a check to see if the popup is closed before making the API call.

Reply:
Prevented multiple popups from being opened.

Reply:
Nice change. Works great!